be/src/storage/rowset/data_sample.cpp

```diff
@@ -0,0 +1,342 @@
+    RETURN_IF(_probability_percent == 0 || _probability_percent == 100,
+              Status::InvalidArgument("percent should be in (0, 100)"));
+    std::mt19937 mt(_random_seed);
+    std::bernoulli_distribution dist(_probability_percent / 100.0);
+
+    size_t sampled_blocks = 0;
+    size_t total_blocks = _total_rows / _rows_per_block;
+    RowIdSparseRange sampled_ranges;
+    for (size_t i = 0; i < _total_rows; i += _rows_per_block) {
+        if (dist(mt)) {
+            sampled_blocks++;
+            sampled_ranges.add(RowIdRange(i, std::min(i + _rows_per_block, _total_rows)));
+        }
+    }
+
+    if (sampled_blocks == 0) {
+        std::uniform_int_distribution<size_t> uniform(0, total_blocks);
+        size_t block = uniform(mt);
+        sampled_blocks++;
+        sampled_ranges.add(RowIdRange(block, std::min(block + _rows_per_block, _total_rows)));
+    }
+
+    stats->sample_size += sampled_blocks;
+    stats->sample_population_size += total_blocks;
+
+    return sampled_ranges;
+}
+
+static int compare_datum(const TypeInfoPtr& type_info, const Datum& lhs, const Datum& rhs) {
+    if (lhs.is_null() && rhs.is_null()) {
+        return 0;
+    }
+    if (lhs.is_null() != rhs.is_null()) {
+        if (lhs.is_null())
+            return 1;
+        else
+            return -1;
+    }
+    return type_info->cmp(lhs, rhs);
+}
+
+static int compare_zonemap(const TypeInfoPtr& type_info, const ZoneMapDetail& a, const ZoneMapDetail& b) {
+    int cmp = compare_datum(type_info, a.min_value(), b.min_value());
+    if (cmp != 0) {
+        return cmp;
+    }
+    return compare_datum(type_info, a.max_value(), b.max_value());
+}
+
+void SortableZoneMap::sort() {
+    if (zonemap.empty()) {
+        return;
+    }
+    auto type_info = get_type_info(type);
+
+    // only sort the page_indices but not zonemap
+    page_indices.resize(zonemap.size());
+    std::iota(page_indices.begin(), page_indices.end(), 0);
+    std::sort(page_indices.begin(), page_indices.end(),
+              [&](size_t lhs, size_t rhs) { return compare_zonemap(type_info, zonemap[lhs], zonemap[rhs]) < 0; });
+}
+
+double SortableZoneMap::width(const ZoneMapDetail& zone) {
+    return width(zone.min_value(), zone.max_value());
+}
+
+bool SortableZoneMap::is_support_data_type(LogicalType type) {
+    // TODO: support decimal type
+    switch (type) {
+    case TYPE_TINYINT:
+    case TYPE_SMALLINT:
+    case TYPE_INT:
+    case TYPE_BIGINT:
+    case TYPE_LARGEINT:
+    case TYPE_UNSIGNED_TINYINT:
+    case TYPE_UNSIGNED_SMALLINT:
+    case TYPE_UNSIGNED_INT:
+    case TYPE_UNSIGNED_BIGINT:
+    case TYPE_FLOAT:
+    case TYPE_DOUBLE:
+    case TYPE_DATE:
+    case TYPE_DATETIME:
+    case TYPE_DECIMAL32:
+    case TYPE_DECIMAL64:
+    case TYPE_DECIMAL128:
+        return true;
+    default:
+        return false;
+    }
+}
+
+double SortableZoneMap::width(const Datum& lhs, const Datum& rhs) {
+    if (lhs.is_null() || rhs.is_null()) {
+        return 0;
```

Comment:
why is 0, it seems that is a open range, it seens that infinity or maximum value is reasonable

Reply:
it depends. 
here the width is used to construct the histogram, the null value cannot be put into the histogram, so i just ignore them via set the width to 0